client/state/route/types.ts

```diff
@@ -0,0 +1,45 @@
+/**
+ * Internal dependencies
+ */
+import { ROUTE_CLEAR_LAST_NON_EDITOR, ROUTE_SET } from 'calypso/state/action-types';
+
+export type Query = {
+	[ key: string ]: string | number | undefined;
+	_timestamp?: number;
+};
+
+export type QueryState = {
+	initial: Query | false;
+	current: Query | false;
+	previous: Query | false;
+};
+
+export type Path = string;
+
+export type PathState = {
+	initial: Path;
+	current: Path;
+	previous: Path;
+};
+
+export type RouteState = {
+	path: PathState;
+	query: QueryState;
+};
+
+type RouteClearAction = {
+	type: typeof ROUTE_CLEAR_LAST_NON_EDITOR;
+};
+
+type RouteSetAction = {
+	path: Path;
+	query: Query;
+	type: typeof ROUTE_SET;
+};
```

Comment:
I think that diligently declaring strict types for the `Route*` actions is not necessary. The types are not actually used anywhere.

The `setRoute` action creator (like any other action creator) returns an opaque action object whose internal structure is not interesting. The action is only dispatched by passing it to `dispatch` and its fields could as well be private.

It's only the reducer that actually reads them, but our reducer is declared as accepting `AnyAction` and doesn't use the specific types.

Because of that, the actions creators could simply return the `Action` type from `redux`.